### PR TITLE
[API revision] Separate TransportAuthenticator and PerRPCCredentials

### PR DESCRIPTION
**Pull request by menghanl**
* PerRPCCredentials is used in each RPC
* TransportAuthenticator is used to setup connection

### CODE REVIEW

**Comment by jboeuf** [comment]
Thanks much for the PR! taking a look now.

File: clientconn.go [hunk 1]
```diff
@@ -170,17 +170,17 @@ func WithInsecure() DialOption {
 
 // WithTransportCredentials returns a DialOption which configures a
 // connection level security credentials (e.g., TLS/SSL).
-func WithTransportCredentials(creds credentials.TransportAuthenticator) DialOption {
+func WithTransportCredentials(auth credentials.TransportAuthenticator) DialOption {
```
Comment (jboeuf) [inline clientconn.go]: it looks a bit weird to me have a function called `withTransportCredentials` that takes a `TransportAuthenticator` object.
In C/C++ we call these `ChannelCredentials` so maybe we could align here with:
```
func WithChannelCredentials(auth credentials.ChannelCredentials) DialOption
```
or just be consistent
```
func WithTransportAuthenticator(auth credentials.TransportAuthenticator) DialOption
```

File: clientconn.go [hunk 2]
```diff
@@ -170,17 +170,17 @@ func WithInsecure() DialOption {
...
 	return func(o *dialOptions) {
-		o.copts.AuthOptions = append(o.copts.AuthOptions, creds)
+		o.copts.Authenticator = auth
 	}
 }
 
 // WithPerRPCCredentials returns a DialOption which sets
 // credentials which will place auth state on each outbound RPC.
-func WithPerRPCCredentials(creds credentials.Credentials) DialOption {
+func WithPerRPCCredentials(creds credentials.PerRPCCredentials) DialOption {
```
Comment (jboeuf) [inline clientconn.go]: In C++ we call these `CallCredentials`. It could be a good idea to align as well.

**Comment by jboeuf** [comment]
I think that that it would also awesome to have a function like this:
```
func WithGoogleDefaultCredentials() DialOption
```
Which would, for now, setup the DialOption with SSL and the Google Default Core credentials (scopeless). Please reach out directly to me for more clarifications.

File: clientconn.go (hunk 1 above)
Comment (menghanl) [inline clientconn.go]: Thanks for your review.
The reason for not change the function name is this is a public api, and we are trying to minimize breakage.

How about changing the interface name from `TransportAuthenticator` to `TransportCredentials`? So this function will look like
```
func WithTransportCredentials(creds credentials.TransportCredentials) DialOption
```

**Comment by iamqizhao** [comment]
@jboeuf Even though we are making API breaking change, we still want to minimize the impact. :) I would propose to
i) change WithTransportCredentials(TransportAuthenticator) to WithTransportCredentials(TransportCredentials);
ii) keep WithPerRPCCredentials(PerRPCCredentials) intact.
In this way, we do not break users unless they implemented their own Credentials/TransportAuthenticator previously and these names sound sane to me. What do you think?

WithGoogleDefaultCredentials is good. It is an incremental change to this PR. Let's do it in a separate PR.

File: clientconn.go [hunk 3]
```diff
@@ -369,17 +369,14 @@ func (cc *ClientConn) newAddrConn(addr Address, skipWait bool) error {
...
-			}
-		}
-		if !ok {
+		if ac.dopts.copts.Authenticator == nil {
 			return errNoTransportSecurity
 		}
 	} else {
-		for _, cd := range ac.dopts.copts.AuthOptions {
+		if ac.dopts.copts.Authenticator != nil {
+			return errCredentialsMisuse
```
Comment (iamqizhao) [inline clientconn.go]: The current errCredentialsMisuse does not match here. You need to revise the error msg and comments.

**Comment by jboeuf** [comment]
@iamqizhao 
SGTM. Thanks!

**Comment by jboeuf** [comment]
And please CC me on the name PR for `WithGoogleDefaultCredentials`. Thanks!

File: credentials/credentials.go [hunk 4]
```diff
@@ -54,9 +54,9 @@ var (
 	alpnProtoStr = []string{"h2"}
 )
 
-// Credentials defines the common interface all supported credentials must
+// PerRPCCredentials defines the common interface all supported per RPC credentials must
```
Comment (iamqizhao) [inline credentials/credentials.go]: PerRPCCredentials defines the common interface for the credentials which need to attach security information to every RPC (e.g., oauth2).

**Pull request merged**